app/models/cluster_group.rb

```diff
@@ -395,39 +395,69 @@ def cell_name_index(study_cells)
   end
 
   # create all necessary data array entries for cell_name_index
-  def create_cell_name_index!
-    return nil unless can_index?
-
-    update!(is_indexing: true)
-    study_cells = study.all_cells_array
-    cell_index = cell_name_index(study_cells)
+  def create_cell_name_index!(study_cells, subsample_annotation: nil, subsample_threshold: nil)
+    cell_index = cell_name_index(study_cells, subsample_annotation:, subsample_threshold:)
     # if some cells are not indexed, don't create array as this will break things downstream
     index_count = cell_index.compact.size
-    if cell_index.empty?
+    expected_cells = subsample_threshold || points
+    if cell_index.empty? && subsample_threshold.nil?
       Rails.logger.info "using default cell index for #{study.accession}:#{name}"
-      update!(is_indexing: false, indexed: true, use_default_index: true)
-    elsif index_count != points
-      Rails.logger.info "aborting cell index for #{study.accession}:#{name} - #{points - index_count} cells not found"
-      update!(is_indexing: false)
+      update!(use_default_index: true)
+    elsif index_count != expected_cells
+      Rails.logger.info "aborting cell index for #{study.accession}:#{name} - #{expected_cells - index_count} cells not found"
     else
       cell_index.each_slice(DataArray::MAX_ENTRIES).with_index do |slice, index|
         begin
           DataArray.create!(
             name: 'index', array_type: 'cells', array_index: index, values: slice, linear_data_type: 'ClusterGroup',
-            linear_data_id: id, study_id: study.id, study_file_id:, cluster_name: name
+            linear_data_id: id, study_id: study.id, study_file_id:, cluster_name: name,
+            subsample_annotation:, subsample_threshold:
           )
         rescue => e
-          update!(is_indexing: false)
-          ErrorTracker.report_exception(e, nil, study, self, { job: :create_cell_name_index! })
+          context = { job: :create_cell_name_index!, subsample_annotation:, subsample_threshold: }
+          ErrorTracker.report_exception(e, nil, study, self, context)
         end
       end
-      update!(is_indexing: false, indexed: true)
     end
   end
 
+  # create all necessary cell index arrays, both full-resolution and subsampled
+  def create_all_cell_indices!
```

Comment:
Beyond cell indexing, do we have other ingest tasks that aren't in Ingest Pipeline?  I don't see it as urgent, but having this processing on the app server strikes me as a lurking bottleneck for parallel ingest.  

As we increase subsampling thresholds for larger studies (e.g. 250K), this limitation seems like it would become more substantial.

I imagine porting this to Ingest Pipeline would be straightforward, not worth doing imminently, but worth keeping in mind.

Reply:
Thats a decent point - I don't see this as being a large issue given the sparsity of study creation, but monitoring the backfill migration will give us a good indication of any bottlenecks.  Last time it rose CPU to ~80% without any other deleterious effects.  Also, while we could definitely do this on an ingest machine, I think it's a lot easier to do it in a background process on the app server since we can leverage classes/models.  But if we see lots of  CPU saturation we can move it off easily enough.